askama/src/filters/humansize.rs

```diff
@@ -43,82 +41,100 @@ impl FastWritable for FilesizeFormatFilter {
         dest: &mut W,
         values: &dyn Values,
     ) -> crate::Result<()> {
-        if self.0 < 1e3 {
+        if self.0 < 1000 {
             (self.0 as u32).write_into(dest, values)?;
             Ok(dest.write_str(" B")?)
-        } else if let Some((prefix, factor)) = SI_PREFIXES
-            .iter()
-            .copied()
-            .find_map(|(prefix_factor, max)| (self.0 < max).then_some(prefix_factor))
-        {
-            // u32 is big enough to hold the number 999_999
-            let scaled = (self.0 * factor) as u32;
-            (scaled / 100).write_into(dest, values)?;
-            format_frac(&mut MaybeUninit::uninit(), prefix, scaled).write_into(dest, values)
         } else {
-            too_big(self.0, dest)
-        }
-    }
-}
+            // find appropriate unit to format in
+            let unit = match SI_PREFIXES.binary_search_by_key(&self.0, |&(_, m)| m) {
+                Ok(idx) => SI_PREFIXES[idx],
+                Err(idx) => SI_PREFIXES[idx.saturating_sub(1)],
+            };
+            // divide number up into "full" units, and remainder
+            let integer = self.0 / unit.1;
+            let remainder = self.0 % unit.1;
 
-/// Formats `buffer` to contain the decimal point, decimal places and unit
-fn format_frac(buffer: &mut MaybeUninit<[AsciiChar; 8]>, prefix: AsciiChar, scaled: u32) -> &str {
-    // LLVM generates better byte code for register sized buffers
-    let buffer = buffer.write(AsciiStr::new_sized("..0 kB"));
-    buffer[4] = prefix;
+            // Compute fractional part with desired precision
+            let precision = self.1;
+            let mut scale = 10u128.pow(precision as u32);
```

Comment:
Why is this needed? Can't the scale be derived by the value found in SI_PREFIXES?

Reply:
The scale handles the `precision`. If you e.g. want 3 decimals, the scale will be `1000`.

Example for a precision of 2 (scale = `100`):
Take for example the number of bytes:  `1024`. The best fitting prefix is `1000` (= kB).
So the `integer` portion will be `1024 / 1000 = 1`, and the remainder will be `24`.
The `fractional` portion is generated by first scaling the remainder up and then dividing by unit: `24 * 100 / 1000 = 2`.

The idea is, that this allows to use integer formatting to print the part after the decimal separator. I can then simply format the two integers: `format!("{integer}.{fractional}")`.

Big advantage of this is, that I can freely change the precision (compared to the original version that was fixed to 2).
Or did I misunderstand your question?